File: Exercise_2.cpp
// Time Complexity                              : O(log n)           
// Space Complexity                             : O(1)
// Did this code successfully run on Leetcode   : Yes
// Any problem you faced while coding this      : No


// Your code here along with comments explaining your approach

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int findMin(vector<int>& nums) {
        int n = nums.size();
        if(n == 1 or nums[0] < nums[n-1]) return nums[0];
        int l = 0, r = n-1, md;
        while(l <= r) {
            md = l + (r-l)/2;
            if(nums[md] > nums[md+1]) break;
            if(nums[md] < nums[0]) r = md - 1;
            else if (nums[md] >= nums[0]) l = md+1;
        }
        return nums[md] > nums[md+1] ? nums[md+1] : nums[md];
    }
};

File: Exercise_2.java
class Solution {
    public int findMin(int[] nums) {
        int s = 0;
        int e = nums.length-1;
        if(nums.length == 0){
            return -1;
        }
        if(nums.length == 1){
            return nums[s];
        }
        while (s<=e){
            int m = s + (e-s)/2;
            if (nums.length>2 && nums[m-1]>nums[m]){
                return nums[m];
            }
            if (nums[s]<=nums[m]){
                if(nums[m]<=nums[e]){
                return nums[s];
            }}
            if (nums[s]>nums[e] && nums[m]>=nums[s]){
                s=m+1;
            }
            else if (nums[s]>nums[e] && nums[m]<nums[s]){
                e=m;
            }
                else {return -1;}
        }
        return -1;
    }
}
